Source/engine/assets.cpp

```diff
@@ -65,50 +67,140 @@ bool OpenMpqFile(const char *filename, MpqArchive **archive, uint32_t *fileNumbe
+		result.directHandle = SDL_RWFromFile(relativePath.c_str(), "rb");
+		if (result.directHandle != nullptr) {
+			return result;
+		}
+	}
 
 	// Files in the `PrefPath()` directory can override MPQ contents.
 	{
 		const std::string path = paths::PrefPath() + relativePath;
-		if ((rwops = OpenOptionalRWops(path)) != nullptr) {
+		result.directHandle = OpenOptionalRWops(path);
+		if (result.directHandle != nullptr) {
 			LogVerbose("Loaded MPQ file override: {}", path);
-			return rwops;
+			return result;
 		}
 	}
 
-	// Load from all the MPQ archives.
-#ifdef UNPACKED_MPQS
-	if ((rwops = OpenUnpackedMpqFile(relativePath)) != nullptr)
-		return rwops;
-#else
-	MpqArchive *archive;
-	uint32_t fileNumber;
-	if (OpenMpqFile(filename, &archive, &fileNumber))
-		return SDL_RWops_FromMpqFile(*archive, fileNumber, filename, threadsafe);
-#endif
+	// Look for the file in all the MPQ archives:
+	if (FindMpqFile(filename, &result.archive, &result.fileNumber)) {
+		result.filename = filename;
+		return result;
+	}
 
 	// Load from the `/assets` directory next to the devilutionx binary.
-	if ((rwops = OpenOptionalRWops(paths::AssetsPath() + relativePath)))
-		return rwops;
+	result.directHandle = OpenOptionalRWops(paths::AssetsPath() + relativePath);
+	if (result.directHandle != nullptr)
+		return result;
 
 #if defined(__ANDROID__) || defined(__APPLE__)
 	// Fall back to the bundled assets on supported systems.
 	// This is handled by SDL when we pass a relative path.
-	if (!paths::AssetsPath().empty() && (rwops = SDL_RWFromFile(relativePath.c_str(), "rb")))
-		return rwops;
+	if (!paths::AssetsPath().empty()) {
+		result.directHandle = SDL_RWFromFile(relativePath.c_str(), "rb");
+		if (result.directHandle != nullptr)
+			return result;
+	}
+#endif
+
+	return result;
+}
 #endif
 
-	return nullptr;
+AssetHandle OpenAsset(AssetRef &&ref, bool threadsafe)
+{
+#if UNPACKED_MPQS
+	return AssetHandle { OpenFile(ref.path, "rb") };
+#else
+	if (ref.archive != nullptr)
+		return AssetHandle { SDL_RWops_FromMpqFile(*ref.archive, ref.fileNumber, ref.filename, threadsafe) };
+	if (ref.directHandle != nullptr) {
+		// Transfer handle ownership:
+		SDL_RWops *handle = ref.directHandle;
+		ref.directHandle = nullptr;
+		return AssetHandle { handle };
+	}
+	return AssetHandle { nullptr };
+#endif
+}
+
+AssetHandle OpenAsset(const char *filename, bool threadsafe)
+{
+	AssetRef ref = FindAsset(filename);
+	if (!ref.ok())
+		return AssetHandle {};
+	return OpenAsset(std::move(ref), threadsafe);
+}
+
+AssetHandle OpenAsset(const char *filename, size_t &fileSize, bool threadsafe)
+{
+	AssetRef ref = FindAsset(filename);
+	if (!ref.ok())
+		return AssetHandle {};
+	fileSize = ref.size();
+	return OpenAsset(std::move(ref), threadsafe);
+}
+
+SDL_RWops *OpenAssetAsSdlRwOps(const char *filename, bool threadsafe)
+{
+#ifdef UNPACKED_MPQS
+	AssetRef ref = FindAsset(filename);
+	if (!ref.ok())
+		return nullptr;
+	return SDL_RWFromFile(ref.path, "rb");
+#else

```

Comment:
Isn't the SDL path only for `#ifndef`?

Reply:
We still use the SDL path for video rendering and music playback, which both accept only SDL_RWops